mllib/src/main/scala/org/apache/spark/ml/feature/LabelIndexer.scala

```diff
@@ -0,0 +1,115 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ml.feature
+
+import org.apache.spark.annotation.AlphaComponent
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.{Estimator, Model}
+import org.apache.spark.sql.DataFrame
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.types.{IntegerType, StringType, StructField, StructType}
+import org.apache.spark.util.collection.OpenHashMap
+
+/**
+ * Base trait for [[LabelIndexer]] and [[LabelIndexerModel]].
+ */
+private[feature] trait LabelIndexerBase extends Params with HasLabelCol with HasOutputCol {
+
+  /** Validates and transforms the input schema. */
+  protected def validateAndTransformSchema(schema: StructType, paramMap: ParamMap): StructType = {
+    val map = this.paramMap ++ paramMap
+    val labelType = schema(map(labelCol)).dataType
+    require(labelType == StringType, s"The label column must be string-typed but got $labelType.")
+    val inputFields = schema.fields
+    val outputColName = map(outputCol)
+    require(inputFields.forall(_.name != outputColName),
+      s"Output column $outputColName already exists.")
+    // TODO: Store ML attributes in metadata.
+    val outputFields = inputFields :+ StructField(map(outputCol), IntegerType, nullable = false)
+    StructType(outputFields)
+  }
+}
+
+/**
+ * :: AlphaComponent ::
+ * A label indexer that maps a string column of labels to an integer column of label indices.
+ * The indices are in [0, numLabels), ordered by label frequencies.
+ * The most frequent label gets index 0.
+ */
+@AlphaComponent
+class LabelIndexer extends Estimator[LabelIndexerModel] with LabelIndexerBase {
+
+  /** @group setParam */
+  def setLabelCol(value: String): this.type = set(labelCol, value)
+
+  /** @group setParam */
+  def setOutputCol(value: String): this.type = set(outputCol, value)
+
+  // TODO: Allow unseen labels in transform.
+
+  override def fit(dataset: DataFrame, paramMap: ParamMap): LabelIndexerModel = {
+    val map = this.paramMap ++ paramMap
+    val counts = dataset.select(map(labelCol)).map(_.getString(0)).countByValue()
+    val labels = counts.toSeq.sortBy(-_._2).map(_._1).toArray
+    val model = new LabelIndexerModel(this, map, labels)
+    Params.inheritValues(map, this, model)
+    model
+  }
+
+  override def transformSchema(schema: StructType, paramMap: ParamMap): StructType = {
+    validateAndTransformSchema(schema, paramMap)
+  }
+}
+
+/**
+ * :: AlphaComponent ::
+ * Model fitted by [[LabelIndexer]].
+ */
+@AlphaComponent
+class LabelIndexerModel private[ml] (
+    override val parent: LabelIndexer,
+    override val fittingParamMap: ParamMap,
+    labels: Array[String]) extends Model[LabelIndexerModel] with LabelIndexerBase {
```

Comment:
The `labels` in a fitted model should be viewed as immutable. If there are new labels, we can train a new `LabelIndexerModel`. We are gonna put label names into the metadata, so we still tracks which index maps to which label.

We can add an option to `LabelIndexer` that allows users to specify part of the mapping. If we implement this option, then we can keep the original ordering unchanged on new datasets.